[ 1팀 ] 시온 - 서버 배포 및 OAuth 리뷰 요청

ehdrhelr opened the pull request:
안녕하십니까, 즐거운 석가탄신일입니다.

지난 프로젝트에서 하지 못했던 망 분리된 서버 배포와 OAuth를 우선 구현했습니다.

### 배포

- VPC를 생성하고 public subnet과 private subnet을 구분하여 웹서버용 ec2는 public subnet, 데이터베이스용 RDS는 private subnet에 생성하였습니다.
- NginX와 톰캣을 연동하여 8080포트는 외부에 개방하지 않고 80포트만 개방했습니다.
- 로컬에서 쉘 스크립트를 이용해 프로젝트를 build하고 생성된 jar파일을 S3 버켓에 업로드합니다.
- 웹서버에서 S3 버켓에 업로드된 jar파일을 다운로드 받아 배포하고 버켓의 jar파일을 삭제하는 스크립트를 crontab으로 매분 동작시켜 배포를 자동화시켰습니다.
- 배포시 webhook를 이용해 슬랙 팀 채널에 알림이 가도록 설정했습니다.

### OAuth

- 학습 차원에서 카카오 로그인을 먼저 구현했습니다.
- 사용자가 카카오 로그인을 하면 콜백 URL로 Authorization code를 받고 이 code를 이용해 access token을 요청하여 얻습니다.
- access token으로 사용자 정보를 불러온 후 세션에 저장하여 로그인 처리합니다.
- 아직 다른 클래스분들과 협의한 부분이 없지만 코드리뷰를 받고자 우선 웹에 구현했습니다.
- 추후 더 학습하여 JWT를 이용해 GitHub 로그인을 구현해 볼 예정입니다. 

### 배포주소 : http://13.125.35.62/

감사합니다.

jung-yun committed:
Update README.md

jung-yun committed:
Create README.md

jung-yun committed:
Create README.md

jung-yun committed:
Create README.md

Commit:
[#1] 프로젝트 초기 세팅

Commit:
[#2] chore : 서버 배포

Commit:
[#3] feature : 카카오 로그인 인증 구현

카카오 API 서버로부터 AccessToken을 발급받을 수 있는 code값을 전달 받을 수 있다. (인증처리)

Commit:
[#3] feature : 카카오 로그인 토큰 발급받기 구현

카카오 API에 code값을 전달하여 AccessToken을 발급 받을 수 있다. (권한부여)

Commit:
[#3] feature : 카카오 로그인 토큰으로 사용자 정보 가져오기 구현

Commit:
[#3] feat : 카카오 로그인 json으로 가져온 사용자 정보를 자바 오브젝트로 변경

User 엔티티 추가

Commit:
[#3] chore : JDBC 의존성 추가하고 MySQL 연결

Commit:
[#3] feat : 카카오 로그인시 회원가입 구현

Commit:
[#3] feat : 카카오 로그인시 가입된 회원인지 확인 후 가입되어 있지 않다면 가입

Commit:
[#3] refactor : 카카오 로그인 메소드 분리와 환경변수 사용, DB 환경변수 사용

Commit:
[#3] refactor : KakaoLoginUtils 구조 변경

Commit:
[#3] chore : handlebars 사용

의존성 추가
uri에 properties 사용

Commit:
[#3] chore : 잘못 업로드된 application-oauth.properties 삭제

Commit:
[#3] feat : 카카오 로그인, 로그아웃 구현

1. 카카오 로그인하면 회원가입하지 않고 로그인 정보 세션에 저장하고 애노테이션으로 로그인상태 확인
2. 초기 화면 수정
3. 리팩토링 요망

Commit:
[#3] refactor : 중복코드 메소드화

Commit:
[#3] chore : 배포스크립트 추가

메소드 위치 변경
빈줄 제거
EOF 개행

Commit:
[#3] style : 불필요한 주석 제거, EOF 개행

Commit:
Merge pull request #14 from ehdrhelr/feature/BE/oauth

[#3] 카카오 로그인 구현

Commit:
chore : 배포스크립트에 슬랙 웹후크 주소 노출되어 삭제함

realsigridjin commented:
지나가다가 카카오 로그인 해봤는데 500 에러 뜹니다 ㅜㅜ
![image](https://user-images.githubusercontent.com/41055141/118777867-80446980-b8c4-11eb-9474-0c6303c024b7.png)

ghost commented:
@jypthemiracle 친절한 에러 확인 감사합니다..! 🥲

카카오로부터 받아온 사용자 정보(Json)를 자바 오브젝트로 변환하는 과정에서 클래스에 누락된 property key가 있어서 오류가 발생했습니다. 

제 아이디로 테스트 할 땐 없었던 property key 두개(properties.profile_image, properties.thumbnail_image)가 시그리드가 로그인을 시도할 때 나타나서 두 key도 받아서 변환할 수 있도록 수정했습니다.

Commit:
[#3] fix : 프사 있는 사용자가 로그인 해도 500에러 뜨지 않도록 수정

카카오로부터 받아온 사용자 정보(Json)를 자바 오브젝트로 변환하는 과정에서 클래스에 누락된 property key 두개 추가
properties.profile_image, properties.thumbnail_image

Commit:
Merge pull request #15 from ehdrhelr/BE/fix/aouth

[#3] fix : 프사 있는 사용자가 로그인 해도 500에러 뜨지 않도록 수정

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/config/auth/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        boolean isLoginUserAnnotation = parameter.getParameterAnnotation(LoginUser.class) != null;
```
더 좋은 메소드가 있는 것 같습니다.

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/MethodParameter.html#hasParameterAnnotation-java.lang.Class-

ehdrhelr replied:
오... 훨씬 깔끔하고 멋진 코드가 되었습니다.

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/controller/UserController.java:
```diff
@@ -0,0 +1,39 @@
+
+@RequiredArgsConstructor
+@Controller
+public class UserController {
```
전반적으로 이 컨트롤러에서 하는 일은 사실 Interceptor나 Filter에서 수행해야 할 일입니다. 
컨트롤러에 정보가 들어오기 전에 처리돼야 할 로직들인데요, 지금 프로젝트에선 어려울 수도 있겠지만 컨트롤러로 요청이 진입하기까지 거치는 과정들에 대해 학습해 보시고, 유저의 로그인/로그아웃 처리는 어떻게 이뤄져야할지도 같이 고민해보시면 좋겠어요.

ehdrhelr replied:
인터셉터를 학습하고 살짝 적용해보았는데요, 현재 프로젝트에선 로그인할 때 현재 로그인 상태확인, 로그아웃할 때 세션 삭제 정도를 인터셉터에서 처리하도록 구현했습니다.
인터셉터에서 userService까지 주입받아 토큰으로 유저정보를 가져오는 로직까지 처리해야하나 고민했봤는데 그러면 컨트롤러 역할이 없는 것 같아서 그냥 두었습니다.
인터셉터에선 로그인/로그아웃시 세션 유효성 검사나 세션을 set 또는 remove해주는 정도로 처리하면 되는지 궁금합니다..!

프로젝트를 진행하면서 인터셉터를 적극적으로 사용해보겠습니다.

wheejuni replied:
@ehdrhelr 인터셉터와 필터에 대해 적극적으로 탐구해보시고 더 나아가 이런 컴포넌트들에서 획득한 사용자 인증 정보를 그 다음 요청 처리 플로우에 보내주려면 어떤 방법을 취해야할지도 학습해보세요. 세션은 무엇인가요? 왜 세션을 생각했죠? 세션 없이는 어려울까요?

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/domain/User.java:
```diff
@@ -0,0 +1,29 @@
+import team01.airbnb.dto.KakaoProfile;
+import team01.airbnb.dto.RoleType;
+
+import java.util.Date;
```
`Date` 는 이제 놓아줍시다.
https://madplay.github.io/post/reasons-why-javas-date-and-calendar-was-bad

ehdrhelr replied:
자꾸 쓰게되는 Date... LocalDate로 바꾸었습니다.

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/dto/KakaoProfile.java:
```diff
@@ -0,0 +1,37 @@
+import lombok.Data;
+
+@Data
+public class KakaoProfile {
```
유저 정보에 대해 유일하게 정의된 타입인데 이 타입이 특정 서비스나 외부 공급자에 강하게 결합되면 안됩니다. 
앞으로 서비스가 대성하면 네이버, 애플, GitHub 등등 많이 붙여야할텐데 그때는 어떤 참사가 일어날까요?
지금 `KakaoProfile` 기반으로 작성한 로직들은 모두 어떻게 되는 걸까요? 

정보의 캡슐화 측면에서도 바라보면 우리가 카카오에서 던져주는 유저 정보를 모두 취해야 하는 것은 아니죠. 
취해야 할 것은 취하고, 나머지는 버려야 할 겁니다. 
물론 JSON 역/직렬화를 해야하다보니 우선 넘어오는 JSON을 다 받긴 해야하지만, _다양한 소셜 로그인 공급사의 유저 프로필 정보에서 우리가 공통적으로 뽑아내야 할 정보는 무엇인가?_ 를 먼저 고민하시기 바랍니다.

ehdrhelr replied:
카카오 이외의 다른 소셜 로그인 서비스가 생긴다면 NaverProfile, GoogleProfile 등등 만들 생각이었는데, 저희 서비스에서 필요한 공통 정보( ex: 이름, 이메일)만 갖는 타입을 만든다면 다른 소셜 로그인 서비스가 추가되어도 저희 서비스에서 정의한 타입으로 처리할 수 있을 것 같습니다.

다만 이 경우 각각의 소셜 로그인 서비스에서 넘어오는 JSON에 차이가 있어 이름과 이메일을 추출하는 로직이 서비스마다 분리될 것 같은 예감이 듭니다. 위의 공통정보만 갖는 타입으로 일치시킨 것처럼 로직 또한 최대한 공통적인 부분을 처리하도록 구성하면 될까요??

wheejuni replied:
우선 공통적으로 사용할 타입을 정의하는 시도를 해보시면 좋을 것 같고,
구체적인 구현체는 당연히 서비스마다 달라질 겁니다. 그러나 가장 중요한 것은 이 도메인 객체에 의존하는 다른 클래스들이 **객체의 구체적인 구현에 대해 전혀 몰라도 상관없도록** 유연하게 구현하는 것입니다.

ehdrhelr replied:
오홍 좋은 인사이트 주셔서 감사합니다!

1. SocialProfile 인터페이스를 만들고 필요한 정보를 가져올 수 있도록 추상메서드를 선언했습니다. (getUsername, getEmail)
2. KakaoProfile 클래스는 SocialProfile을 구현하고 KakaoProfile에 의존하고 있던 로직을 모두 SocialProfile에 의존하도록 수정했습니다.
3. 소셜 로그인 서비스에 엑세스 토큰 및 프로필 정보를 요청하는 메서드 또한 uri 및 requestEntity를 파라미터로 전달해 어느 소셜 로그인이든 사용하게 하고 싶었으나 아직 다른 소셜 로그인 서비스에 대한 학습이 되지 않아 수정하지 못했습니다.

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/dto/KakaoProfile.java:
```diff
@@ -0,0 +1,37 @@
+    public Integer id;
+    public String connected_at;
+    public Properties properties;
+    public KakaoAccount kakao_account;
```
`@JsonProperty` 애노테이션을 사용하여 자바의 코딩 컨벤션을 항상 준수하는 코딩을 해 주시면 좋겠네요.
그리고 꼭 `public` 이여야 하나요?

ehdrhelr replied:
핑계지만 json데이터를 자바 오브젝트로 변경해주는 사이트를 이용만하고 이후에는 신경을 안썼네요 ㅠㅠㅠ
수정했습니다..!

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/service/UserService.java:
```diff
@@ -0,0 +1,104 @@
+                user.setCreateDate(rs.getDate("join_date"));
+                return user;
+            });
+        } catch (EmptyResultDataAccessException e) {
+            return null;
+        }
```
`return null` 을 하면 애써 `try - catch` 를 한 의미가 없어집니다.
대개 이런 경우 정석적인 처리는 checked exception을 래핑하여 프로젝트에서 사용하는 runtime exception으로 바꿔 던지는 처리 기법을 쓰는 것입니다.
이렇게 런타임 예외로 바꾸면 그때부턴 스프링이 제공하는 예외 처리 플로우를 태울 수 있고, 표준화된 예외 처리가 가능해집니다. 
`return null` 을 하면 이 메소드에 의존하고 있는 다른 소비자들이, `User` 객체에 액세스하는 순간 NPE가 발생하겠죠.
이런 무책임한 코드... 괜찮을까요?

ehdrhelr replied:
용납할 수 없죠...ㅜㅜ 그동안 예외처리를 그다지 신경쓰지 않았는데 관심을 많이 가져야할 것 같습니다..!

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/service/UserService.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Transactional
+    public int join(User user) {
```
여기서 `int` 를 리턴하는 것의 의미를 찾기 어려울 것 같네요.
이 메소드의 최종 소비자가 어떤 결과값을 기대할지 생각해보시면 바로 답이 나올 것 같아요.
(아마 코딩하시면서 느끼기도 하셨을 겁니다.)

이 메소드를 호출한 측은 회원 가입 로직을 수행하기 위해 수행한 것이고, 그의 관심사는 회원 가입 성공 여부일 것입니다.
그렇다면 여기서 정수형 리턴 타입을 가지는 것 외에... 더 좋은 방법이 있을 것 같네요.

ehdrhelr replied:
쿼리 실행 결과를 그대로 리턴해주려고 했는데, 그것보다 최종소비자가 기대하는 결과 값을 생각해서 boolean으로 리턴해주는 방법이 더 좋을 것 같습니다..!

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/service/UserService.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private <T> T convertJsonToObject(ResponseEntity<String> response, Class<T> valueType) {
+        ObjectMapper objectMapper = new ObjectMapper();
```
`ObjectMapper` 는 thread-safe 합니다. 
심지어 스프링의 App Context가 구동될 때 ObjectMapper 타입의 빈이 하나 같이 뜨기도 합니다.
각 컴포넌트에서 주입받아 사용해도 됩니다.

ehdrhelr replied:
짱입니다... 빈으로 또 뭐가 등록되어 있는지 관심이 가네요!

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/service/UserService.java:
```diff
@@ -0,0 +1,104 @@
+        ResponseEntity<String> response = getJsonResponseByPost(
+                kakaoLoginUtils.getProfileUri()
+                , kakaoLoginUtils.getKakaoProfileRequestEntity(accessToken));
+        System.out.println(response.getBody());
```
`System.out` 사용하지 않는 것으로 꾸준히 피드백 드리고 있습니다.

ehdrhelr replied:
Logger로 바꿨습니다 ㅠㅠㅠ

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/service/UserService.java:
```diff
@@ -0,0 +1,104 @@
+        return null;
+    }
+
+    private ResponseEntity<String> getJsonResponseByPost(String uri, @Nullable HttpEntity<?> requestEntity) {
```
타입 파라메터를 `String` 말고 줄 수 있는 방법은 없으려나요?

wheejuni replied:
API 통신을 한 후 결과물을 문자열로 받아, 다시 `ObjectMapper` 에 넣고 역직렬화하는 구현을 하신 것을 보고 드린 피드백이었는데요,
`RestTemplate` 에서 바로 모델 클래스로 받아올 수 있는 방법은 없을까 하는 피드백이었습니다.

ehdrhelr replied:
postByObject를 사용하면 API 통신 후 받은 JSON을 역직렬화 할 필요없이 모델 클래스를 바로 받아올 수 있군요!
감동입니다...🥲

wheejuni commented on BE/airbnb/src/main/java/team01/airbnb/utils/KakaoLoginUtils.java:
```diff
@@ -0,0 +1,73 @@
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+@PropertySource("classpath:application-oauth.properties")
```
스프링 애플리케이션에서 설정 파일을 관리하는 표준 컨벤션에 맞지 않습니다.
만약 이대로라면 _oauth_ 라는 프로필 설정으로 애플리케이션을 구동할 때 읽어들일 설정 파일이 될 것입니다.
앞의 `application-` 을 떼거나, 아예 다른 파일명으로 바꾸는 작업이 필요해 보이네요.

ehdrhelr replied:
오오 그렇군요! 아예 다른 파일명으로 바꿨습니다.
컨벤션을 잘 모르지만 카카오 로그인에 필요한 정보들이 담겨져 있으니 나중에 깃헙 로그인에 필요한 정보 등과 함께 `application-oauth.properties`란 이름으로 사용해도 괜찮지 않을까란 생각이 들었습니다. 현재 카카오 로그인에 필요한 정보가 포함된 oauth라는 프로필 설정으로, 애플리케이션이 구동될 때 읽어들일 설정 파일이 된다면 예상되는 문제가 무엇이 있을지 궁금합니다.

wheejuni replied:
구동할때마다 `oauth` 라는 프로필을 같이 부여할거라면 상관 없습니다. 다만 관리포인트가 하나 더 늘어나는 것이기때문에 자주 쓰지 않는 방법입니다.
개인적으론 지금 의존하고 있는 설정값들이 별도의 파일로 관리될 필요성이 있는지는 모르겠습니다. `application-local.properties`, `application-production.properties` 와 같이 다른 일반 설정값들과 같은 파일에 담겨도 상관 없지 않을까요?

그리고 뭔가 스프링 프로필 관리에 대해 이해가 부족하시다는 느낌도 듭니다. 
https://kogle.tistory.com/220 를 읽어보시고, 다른 검색도 해보시는 것도 좋겠습니다.

ehdrhelr replied:
답변 감사합니다.
스프링 프로필 관리에 대해 더 학습해보고 프로젝트 진행하면서 말씀하신 방향으로 적용해보겠습니다!

wheejuni reviewed (changes requested):
수고 많으셨습니다. 👍 
시작 단계의 코드인만큼 아직 수정해야 할 점도 많지만, 일찍 PR을 오픈해 피드백을 얻고자 하신 점은 감사하기도 하고 좋다고 느껴지네요. 
다만 **자바의 코딩 컨벤션에 대해 늘 의식하고 준수해 주시기를 부탁합니다.**

피드백 남겨드리니 확인해 주세요. 💯

Commit:
[#3] refactor : 꺼내지 않고 물어본다. get() != null -> has()

getParameterAnnotation -> hasParameterAnnotation

Commit:
[#3] refactor : Date는 이제 놓아주자. Date -> LocalDate

Commit:
[#3] refactor : 자바컨벤션을 준수를 위해 변수명 카멜케이스로 변경후 JsonProperty 어노테이션으로 컬럼 지정

Commit:
[#3] refactor : KakaoProfile 내부 접근제어자 변경. public -> private

Commit:
[#3] refactor : 회원가입 성공 여부 boolean return 하도록 변경

Commit:
[#3] refactor : bean으로 등록되어 있는 objectMapper 주입 받아 사용

Commit:
[#3] refactor : System.out -> Logger

Commit:
[#3] style : application-oauth.properties 이름 변경 -> custom-oauth.properties

Commit:
[#3] refactor : 타입 파라미터 변경

Commit:
[#3] chore : assertJ 의존성 추가

Commit:
[#3] refactor : catch문에서 무책임한 return null 제거

runtime exception 발생시켜 던지고 글로벌 예외 핸들러에서 처리할 수 있도록 조치

Commit:
[#3] refactor : SocialProfile 추가하여 다른 소셜 로그인 서비스가 추가되어도 확장에 유연하도록 수정

Commit:
[#3] refactor : 로그인/로그아웃에 Interceptor 살짝 적용

Commit:
Merge pull request #17 from ehdrhelr/BE/refactor/oauth

코드리뷰를 토대로 코드를 개선했습니다.

ghost commented:
친절한 코드리뷰 감사합니다.

제가 기대했던 것보다 훨씬 많은 것을 배울 수 있었던 코드리뷰여서 이렇게 답장드리는게 몹시 기쁘고 뿌듯합니다.
브라이언의 의도대로 코드를 개선하고자 부족하지만 고민을 많이 해봤는데 코드에 제대로 반영되었는지 궁금합니다.
더 개선할 점이 있다면 언제든지 환영입니다.

좋은 주말 되세요!

wheejuni reviewed (approved):
이정도면 우선 남은 프로젝트를 진행하기에 충분한 수준인 것 같습니다.
수고 많으셨고, 앞으로 남은 프로젝트 기간도 더 수고해주세요. 감사합니다. 💯
